ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpInterceptor.java

```diff
@@ -0,0 +1,7 @@
+package io.ripc.protocol.tcp;
+
+public interface TcpInterceptor<I, O, II, OO> {
+
+    TcpHandler<II, OO> intercept(TcpHandler<I, O> handler);
```

Comment:
> Reconnection (and connection management in general) is an important use case we have that currently powers the Spring STOMP support.

Actually, it is a very mainstream usecase in almost all applications. The way we do it in Netflix is by introducing a loadbalancer that does the determination of how to retry/load balance. Since, I have not even introduced the client APIs yet, I would like to keep this discussion for the client APIs. The API here is purely for the server.

>> I think there's also the question of "what exactly are we intercepting" and "to what end?"

In this particular case (TCP server interceptors) we are intercepting "connection processing" that will include but not limited to:

- Modifying input/output.
- Introspecting input/output from the point of view of insights.
- Short-circuiting connection processing i.e. the connection does not even reach the actual handler.
- Metrics gathering of connection processing times.
- Logging

>> I'm wondering if "intercepting" shouldn't be done at the site of the event (Subscription.request and onNext) and as a translation/codec/whatever rather than incurring the overhead of wrapping, delegates, etc...

I think you are looking at it mostly from the point of view reading/writing data, in which case it may be ok to intercept read/write separately. However, that approach does not work well when we start talking about short-circuiting. The current approach of treating server as a function (issue #15) covers all the usecases. I would like to see if you have a usecase which can not be covered by that design.

>> This allows for different interceptors (and even multiple interceptors) on the read and write

The current design does not restrict the server to a single interceptor. I have described above the issue with designing interceptors as being applied on read and write.